vault-tool/src/Network/VaultTool/KeyValueV2.hs

```diff
@@ -0,0 +1,237 @@
+import Data.Time (UTCTime)
+
+import Network.VaultTool.Internal (
+    newDeleteRequest,
+    newGetRequest,
+    newListRequest,
+    newPostRequest,
+    runVaultRequestAuthenticated,
+    runVaultRequestAuthenticated_,
+    withStatusCodes,
+ )
+import Network.VaultTool.Types (
+    VaultConnection,
+    Authenticated,
+    VaultMountedPath (..),
+    VaultSearchPath (..),
+    VaultSecretPath (..),
+ )
+
+data VaultSecretMetadata = VaultSecretMetadata
+    { _VaultSecretMetadata_leaseDuration :: Int
+    , _VaultSecretMetadata_leaseId :: Text
+    , _VauleSecretMetadata_renewable :: Bool
+    }
+    deriving (Show, Eq {- TODO Ord -})
+
+instance FromJSON VaultSecretMetadata where
+    parseJSON = withObject "VaultSecretMetadata" $ \v ->
+        VaultSecretMetadata
+            <$> v .: "lease_duration"
+            <*> v .: "lease_id"
+            <*> v .: "renewable"
+
+data VaultSecretVersion a = VaultSecretVersion
+    { vsvData :: a
+    , vsvMetadata :: VaultSecretVersionMetadata
+    }
+    deriving (Show)
+
+instance FromJSON a => FromJSON (VaultSecretVersion a) where
+    parseJSON = withObject "VaultSecretVersion" $ \v ->
+        VaultSecretVersion
+            <$> v .: "data"
+            <*> v .: "metadata"
+
+data VaultSecretVersionMetadata = VaultSecretVersionMetadata
+    { vsvmCreatedTime :: UTCTime
+    , vsvmDeletionTime :: Maybe UTCTime
+    , vsvmDestroyed :: Bool
+    , vsvmVersion :: Int
+    }
+    deriving (Show)
+
+instance FromJSON VaultSecretVersionMetadata where
+    parseJSON = withObject "VaultSecretVersionMetadata" $ \v ->
+        VaultSecretVersionMetadata
+            <$> v .: "created_time"
+            <*> (v .: "deletion_time" >>= parseOptionalDate)
+            <*> v .: "destroyed"
+            <*> v .: "version"
+      where
+        parseOptionalDate x = parseJSON x <|> pure Nothing
+
+vaultRead ::
+    FromJSON a =>
+    VaultConnection Authenticated ->
+    VaultSecretPath ->
+    IO (VaultSecretVersion a)
+vaultRead conn path = vaultReadVersion conn path Nothing
+
+vaultReadVersion ::
+    FromJSON a =>
+    VaultConnection Authenticated ->
+    VaultSecretPath ->
+    Maybe Int ->
+    IO (VaultSecretVersion a)
+vaultReadVersion conn (VaultSecretPath (mountedPath, searchPath)) version =
+    runVaultRequestAuthenticated conn (newGetRequest path) >>= \(DataWrapper x) -> pure x
+  where
+    path = vaultActionPath ReadSecretVersion mountedPath searchPath <> queryParams
+    queryParams = case version of
+        Nothing -> ""
+        Just n -> "?version=" <> T.pack (show n)
+
+newtype DataWrapper a = DataWrapper a
+
+instance ToJSON a => ToJSON (DataWrapper a) where
+    toJSON (DataWrapper x) = object ["data" .= x]
+
+instance FromJSON a => FromJSON (DataWrapper a) where
+    parseJSON = withObject "DataWrapper" $ fmap DataWrapper . (.: "data")
+
+{- | <https://www.vaultproject.io/docs/secrets/generic/index.html>
+
+ The value that you give must encode as a JSON object
```

Comment:
Is the a hard requirement of Vault?

Reply:
I believe vault allows for arbitrary data so this must be a constraint added by this library.

Reply:
It looks like your type can handle arbitrary data.  Should we remove this comment?

Reply:
2c42064fa561f294710ccd8b3ca290ace3d42d3f